### PR TITLE
MAP-1460 PrisonApiClient getMovement EmptyList When Server Error 404

### PR DESCRIPTION
**Pull request by marcusaleman**
As I am looking to cover webClient line which might be due 4xx server error, 
image

As It's a Reactive Chain response we should handle the error however this mean logic change where we have to use **onErrorResume**

As Consumer might consume 500 server error I have capture that scenario where  Client gives 400 but there is a suppressor in the code returning 500.

plus a EmptyList return not already tested.

**Resources**
PRISON API SWAGGER https://prison-api-dev.prison.service.justice.gov.uk/swagger-ui/index.html#/

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/PrisonApiClient.kt [hunk 1]
```diff
@@ -300,10 +300,13 @@ class PrisonApiClient(
...
-          "${toDate.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)}",
+          toDate.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME),
       )
       .header("Page-Limit", "10000")
       .retrieve()
       .bodyToMono(typeReference<List<Movement>>())
+      .onErrorResume(WebClientResponseException::class.java) {
+        Mono.just(emptyList())
+      }
       .block() ?: emptyList()
```
Comment (marcusaleman) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/PrisonApiClient.kt:311]: Either we trigger a solely WebClientResponseException and then we handle it 4xx server error  , 404 etc.
Or we deliver an emptyList()

But i think original code doesn't deliver 404 when it gets 404 from Prison-API.

Comment (michalnawrockidj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/PrisonApiClient.kt:311]: The code will return 500 error 
Screenshot 2024-09-04 at 13 05 27 .
I do not recommend returning the empty list as this means it is no record for prison XXX. But the reality is the prison XXX not exist. 
If you want to return correct error code I recon to follow this example https://github.com/ministryofjustice/hmpps-welcome-people-into-prison-api/blob/04e77e08aaf4269588aae3656bd61cd9027ae6c3/src/main/kotlin/uk/gov/justice/digital/hmpps/bodyscan/apiclient/BodyScanPrisonApiClient.kt#L41  
It is also good to check in all clients consuming `https://welcome-api-dev.prison.service.justice.gov.uk/prisons/LDF/recent-arrivals` how the 500 and 404 error is communicated to the users.

Comment (marcusaleman) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/PrisonApiClient.kt:312]: @michalnawrockidj 
this was the PRISON-API we tested yesterday,
agencyID = XXX 
{
    "status": 404,
    "userMessage": "Resource with id [XXX] not found.",
    "developerMessage": "Resource with id [XXX] not found."
}
So the code is catching 404 and _I have swap code from emptyList to trigger 404 error.

Comment (marcusaleman) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/prison/PrisonApiClient.kt:312]: After discussion/chat with @michalnawrockidj 
A) As WPIP-UI (Front End) and some consumers might use 500 and We discuss couple of test on the top.
B) I will remove throwing 404 and I will capture the scenario discussed previously

**Comment by marcusaleman** [comment]
I create a new pull request with a clean commit.

**Pull request closed without merging**